### PR TITLE
v2: ucm-validator: add config files for DB845c and Lenovo-YOGA-C630-13Q50

### PR DESCRIPTION
**Pull request by Srinivas-Kandagatla**
Add Config Files support for DB845c and Lenovo-YOGA-C630-13Q50
Changes since v1:
Updated with configs with outputs of alsa-info.sh

### CODE REVIEW

File: python/ucm-validator/configs/misc/Lenovo-YOGA-C630-13Q50.txt [hunk 1]
```diff
@@ -0,0 +1,16142 @@
...
+
+Pulseaudio:
+      Installed - Yes (/usr/bin/pulseaudio)
+      Running - Yes
+
+
+!!Soundcards recognised by ALSA
+!!-----------------------------
+
+ 0 [LenovoYOGAC6301]: Lenovo-YOGA-C63 - Lenovo-YOGA-C630-13Q50
```
Comment (perexg) [inline python/ucm-validator/configs/misc/Lenovo-YOGA-C630-13Q50.txt:65]: Uff. Driver name, card name everything is similar...

Can we join things for the same "controller" and let UCM to structure things on it's own ?

````
diff --git a/sound/soc/qcom/sdm845.c b/sound/soc/qcom/sdm845.c
index b2de65c7f95c..dd6f0f96d4e8 100644
--- a/sound/soc/qcom/sdm845.c
+++ b/sound/soc/qcom/sdm845.c
@@ -557,6 +557,7 @@ static int sdm845_snd_platform_probe(struct platform_device *pdev)
        card->dapm_widgets = sdm845_snd_widgets;
        card->num_dapm_widgets = ARRAY_SIZE(sdm845_snd_widgets);
        card->dev = dev;
+       card->driver_name = "SDM845";
        dev_set_drvdata(dev, card);
        ret = qcom_snd_parse_of(card);
        if (ret) {
````

Comment (Srinivas-Kandagatla) [reply python/ucm-validator/configs/misc/Lenovo-YOGA-C630-13Q50.txt:65]: Thanks for the feedback,

With this proposed change alsa-info.sh shows
!!Soundcards recognised by ALSA
!!-----------------------------

 0 [LenovoYOGAC6301]: SDM845 -  Lenovo-YOGA-C630-13Q50
                                      LENOVO-81JL-LenovoYOGAC630_13Q50-LNVNB161216

and
 pulseaudio is looking for ucm files at locations.
 /usr/share/alsa/ucm2/SDM845/SDM845.conf
 /usr/share/alsa/ucm2/Lenovo-YOGA-C630-13Q50/ Lenovo-YOGA-C630-13Q50.conf

Am not sure if this is what is expected?
I was more of thinking that this proposed change would cleanup the ucm structure and look for this ucm 
 /usr/share/alsa/ucm2/SDM845/Lenovo-YOGA-C630-13Q50.conf 

Am I missing anything here?
thanks,
srini

Comment (perexg) [reply python/ucm-validator/configs/misc/Lenovo-YOGA-C630-13Q50.txt:65]: Do you use the latest alsa-lib (v1.2.2?). This commit fixed the behaviour:  https://github.com/alsa-project/alsa-lib/commit/71a1367bcabc50f99302d8c76395f1cb84975775

Comment (Srinivas-Kandagatla) [reply python/ucm-validator/configs/misc/Lenovo-YOGA-C630-13Q50.txt:65]: from alsa-info I seems be on 1.2.2.

!!ALSA Version
!!------------

Driver version:     k5.7.0-rc1-43712-ge8618528324-dirty
Library version:    1.2.2
Utilities version:  1.2.2

Comment (perexg) [reply python/ucm-validator/configs/misc/Lenovo-YOGA-C630-13Q50.txt:65]: What's strace of 'alsa-ucm -c hw:0 set _verb HiFi' ? It looks correct for HDA-Intel driver:

````
 1 [PCH            ]: HDA-Intel - HDA Intel PCH
                      HDA Intel PCH at 0xd5840000 irq 145
````

````
access("/usr/share/alsa/ucm2/HDA-Intel/HDA Intel PCH at 0xd5840000 irq 145.conf", R_OK) = -1 ENOENT (Adresář nebo soubor neexistuje)
access("/usr/share/alsa/ucm/HDA Intel PCH at 0xd5840000 irq 145/HDA Intel PCH at 0xd5840000 irq 145.conf", R_OK) = -1 ENOENT (Adresář nebo soubor neexistuje)
access("/usr/share/alsa/ucm2/HDA-Intel/HDA-Intel.conf", R_OK) = -1 ENOENT (Adresář nebo soubor neexistuje)
access("/usr/share/alsa/ucm/HDA-Intel/HDA-Intel.conf", R_OK) = -1 ENOENT (Adresář nebo soubor neexistuje)
openat(AT_FDCWD, "/usr/share/alsa/ucm2/HDA-Intel/HDA-Intel.conf", O_RDONLY) = -1 ENOENT (Adresář nebo soubor neexistuje)
````

Comment (Srinivas-Kandagatla) [reply python/ucm-validator/configs/misc/Lenovo-YOGA-C630-13Q50.txt:65]: alsucm seems to be doing right thing here.
faccessat(AT_FDCWD, "/usr/share/alsa/ucm2/SDM845/LENOVO-81JL-LenovoYOGAC630_13Q50-LNVNB161216.conf", R_OK) = -1 ENOENT (No such file or directory)
faccessat(AT_FDCWD, "/usr/share/alsa/ucm/LENOVO-81JL-LenovoYOGAC630_13Q50-LNVNB161216/LENOVO-81JL-LenovoYOGAC630_13Q50-LNVNB161216.conf", R_OK) = -1 ENOENT (No such file or directory)
faccessat(AT_FDCWD, "/usr/share/alsa/ucm2/SDM845/SDM845.conf", R_OK) = -1 ENOENT (No such file or directory)
faccessat(AT_FDCWD, "/usr/share/alsa/ucm/SDM845/SDM845.conf", R_OK) = -1 ENOENT (No such file or directory)
openat(AT_FDCWD, "/usr/share/alsa/ucm2/SDM845/SDM845.conf", O_RDONLY) = -1 ENOENT (No such file or directory)

But Pulse audio seems to be doing something different:

I: [pulseaudio] (alsa-lib)utils.c: could not open configuration file /usr/share/alsa/ucm2/SDM845/SDM845.conf
I: [pulseaudio] (alsa-lib)parser.c: error: could not parse configuration for card SDM845
I: [pulseaudio] (alsa-lib)main.c: error: failed to import hw:0 use case configuration -2
I: [pulseaudio] (alsa-lib)utils.c: could not open configuration file /usr/share/alsa/ucm2/Lenovo-YOGA-C630-13Q50/Lenovo-YOGA-C630-13Q50.conf
I: [pulseaudio] (alsa-lib)parser.c: error: could not parse configuration for card Lenovo-YOGA-C630-13Q50
I: [pulseaudio] (alsa-lib)main.c: error: failed to import Lenovo-YOGA-C630-13Q50 use case configuration -2
I: [pulseaudio] alsa-ucm.c: UCM not available for card Lenovo-YOGA-C630-13Q50

thanks,
srini

Comment (perexg) [reply python/ucm-validator/configs/misc/Lenovo-YOGA-C630-13Q50.txt:65]: Check if PA is not linked to another (older) alsa-lib library binary.

Comment (Srinivas-Kandagatla) [reply python/ucm-validator/configs/misc/Lenovo-YOGA-C630-13Q50.txt:65]: Thanks, Will debug and let you know! Are you happy to send the patch that you proposed, if that is the right thing to do?
Thanks,
srini

Comment (perexg) [reply python/ucm-validator/configs/misc/Lenovo-YOGA-C630-13Q50.txt:65]: Also, PA has another fallback to the card name: https://gitlab.freedesktop.org/pulseaudio/pulseaudio/-/blob/master/src/modules/alsa/alsa-ucm.c#L742 , so the first hw:0 error is correct (and it should be same like alsaucm does). You don't have the config files on the path, so you cannot test.

Comment (Srinivas-Kandagatla) [reply python/ucm-validator/configs/misc/Lenovo-YOGA-C630-13Q50.txt:65]: Ah.. that explains. I was just trying to check the paths!
Previously and now I seems to be always hitting the fallback path..
thanks for explaining this!
Will rearrange configs as below and send new PR.
ucm2/SDM45/LENOVO-81JL-LenovoYOGAC630_13Q50-LNVNB161216.conf
ucm2/SDM45/DB845.conf
--srini

Comment (Srinivas-Kandagatla) [reply python/ucm-validator/configs/misc/Lenovo-YOGA-C630-13Q50.txt:65]: One more question w.r.t ucm2, For some reason if I use ${CardID} in my ucm files, alsaucm and pulseaudio fails to find the card. It complains

"main.c: unable to open ctl device 'hw:${CardId}'

Any pointers to fix this?

Comment (perexg) [reply python/ucm-validator/configs/misc/Lenovo-YOGA-C630-13Q50.txt:65]: Remove cdev lines. The CTL device is already preset for ucm2. And the problem is that the ${CardId} in enable/disable sequences is not substituted (it works only in conditions (If) and Value {} blocks).

**Pull request closed without merging**